File: runtime/lua/_meta/canvas.lua
---@meta _
error("Cannot require a meta file")

lviz.canvas = {}

---@class lviz.canvas.Object
lviz.canvas.Object = {}

---
---@return integer
function lviz.canvas.Object:GetRefCount() end

---@class lviz.canvas.Presentable : lviz.canvas.Object
lviz.canvas.Presentable = {}

---@class lviz.canvas.Geometry : lviz.canvas.Presentable
lviz.canvas.Geometry = {}

---@class lviz.canvas.Point : lviz.canvas.Geometry
---@overload fun(coord: lviz.gp.Pnt):lviz.canvas.Point
---@overload fun(x: number, y: number, z: number):lviz.canvas.Point
lviz.canvas.Point = {}

---@class lviz.canvas.Curve : lviz.canvas.Geometry
lviz.canvas.Curve = {}

---@class lviz.canvas.Circle : lviz.canvas.Curve
---@overload fun(position: lviz.gp.Ax2, radius: number):lviz.canvas.Circle
lviz.canvas.Circle = {}

---@class lviz.canvas.Ellipse : lviz.canvas.Curve
---@overload fun(position: lviz.gp.Ax2, major_radius: number, minor_radius: number):lviz.canvas.Ellipse
lviz.canvas.Ellipse = {}

---@class lviz.canvas.Line : lviz.canvas.Curve
---@overload fun(point1: lviz.gp.Pnt, point2: lviz.gp.Pnt):lviz.canvas.Line
lviz.canvas.Line = {}

---@class lviz.canvas.Polyline : lviz.canvas.Curve
---@overload fun(vertices: lviz.gp.Pnt[]):lviz.canvas.Polyline
lviz.canvas.Polyline = {}

---@class lviz.canvas.TrimmedCurve : lviz.canvas.Curve
---@overload fun(curve: lviz.canvas.Curve, t0: number, t1: number):lviz.canvas.TrimmedCurve
lviz.canvas.TrimmedCurve = {}

---@class lviz.canvas.Surface : lviz.canvas.Geometry
lviz.canvas.Surface = {}

---@class lviz.canvas.Triangle : lviz.canvas.Surface
---@overload fun(point1: lviz.gp.Pnt, point2: lviz.gp.Pnt, point3: lviz.gp.Pnt):lviz.canvas.Triangle
---@overload fun(point1: lviz.gp.Pnt, point2: lviz.gp.Pnt, point3: lviz.gp.Pnt, normal1: lviz.gp.Vec, normal2: lviz.gp.Vec, normal3: lviz.gp.Vec):lviz.canvas.Triangle
lviz.canvas.Triangle = {}

File: src/bind/canvas.cpp
#include "bind.h"

#include "../canvas/circle.h"
#include "../canvas/ellipse.h"
#include "../canvas/line.h"
#include "../canvas/point.h"
#include "../canvas/polyline.h"
#include "../canvas/triangle.h"
#include "../canvas/trimmed_curve.h"

void lviz::bind::BindCanvas(lua_State *L) {
  luabridge::getGlobalNamespace(L)
      .beginNamespace("lviz")
      .beginNamespace("canvas")
      .beginClass<canvas::Object>("Object")
      .addFunction("GetRefCount", &canvas::Object::GetRefCount)
      .endClass()
      .deriveClass<canvas::Presentable, canvas::Object>("Presentable")
      .endClass()
      .deriveClass<canvas::Geometry, canvas::Presentable>("Geometry")
      .endClass()
      .deriveClass<canvas::Point, canvas::Geometry>("Point")
      .addConstructorFrom<canvas::handle<canvas::Point>, void(const gp::Pnt &),
                          void(glm::f32, glm::f32, glm::f32)>()
      .endClass()
      .deriveClass<canvas::Curve, canvas::Geometry>("Curve")
      .endClass()
      .deriveClass<canvas::Circle, canvas::Curve>("Circle")
      .addConstructorFrom<canvas::handle<canvas::Circle>,
                          void(const gp::Ax2 &, glm::f32)>()
      .endClass()
      .deriveClass<canvas::Ellipse, canvas::Curve>("Ellipse")
      .addConstructorFrom<canvas::handle<canvas::Ellipse>,
                          void(const gp::Ax2 &, glm::f32, glm::f32)>()
      .endClass()
      .deriveClass<canvas::Line, canvas::Curve>("Line")
      .addConstructorFrom<canvas::handle<canvas::Line>,
                          void(const gp::Pnt &, const gp::Pnt &)>()
      .endClass()
      .deriveClass<canvas::Polyline, canvas::Curve>("Polyline")
      .addConstructorFrom<canvas::handle<canvas::Polyline>,
                          void(const std::vector<gp::Pnt> &)>()
      .endClass()
      .deriveClass<canvas::TrimmedCurve, canvas::Curve>("TrimmedCurve")
      .addConstructorFrom<canvas::handle<canvas::TrimmedCurve>,
                          void(const canvas::handle<canvas::Curve> &, glm::f32,
                               glm::f32)>()
      .endClass()
      .deriveClass<canvas::Surface, canvas::Geometry>("Surface")
      .endClass()
      .deriveClass<canvas::Triangle, canvas::Surface>("Triangle")
      .addConstructorFrom<
          canvas::handle<canvas::Triangle>,
          void(const gp::Pnt &, const gp::Pnt &, const gp::Pnt &),
          void(const gp::Pnt &, const gp::Pnt &, const gp::Pnt &,
               const gp::Vec &, const gp::Vec &, const gp::Vec &)>()
      .endClass()
      .endNamespace()
      .endNamespace();
}
